File: backend-mock/src/routes/bin-packing.js
const express = require('express');
const { loadLabels } = require('../data');
const { ffd }        = require('../ffd');

const router = express.Router();

router.post('/', (req, res) => {
  const { truck_capacity_tons, filters = {}, max_iterations } = req.body || {};
  if (!truck_capacity_tons || truck_capacity_tons <= 0) {
    return res.status(400).json({ error: 'truck_capacity_tons is required and must be positive' });
  }

  let labels = loadLabels().filter(
    l => l.status === 'available_in_stock' || l.status === 'reserved'
  );

  if (filters.warehouse_code) {
    labels = labels.filter(l => l.warehouse_code === filters.warehouse_code);
  }
  if (filters.customer) {
    labels = labels.filter(l => l.customer === filters.customer);
  }

  const plans = ffd(labels, truck_capacity_tons, max_iterations || 1000);
  res.json(plans);
});

module.exports = router;

File: backend/src/api/v1/bin_packing.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional
from ...data.xlsx_loader import load_labels
from ...services.ffd import ffd

router = APIRouter()


class BinPackingFilters(BaseModel):
    warehouse_code: Optional[str] = None
    customer: Optional[str] = None


class BinPackingRequest(BaseModel):
    truck_capacity_tons: float
    filters: Optional[BinPackingFilters] = None
    max_iterations: int = 1000


@router.post("/")
def run_bin_packing(body: BinPackingRequest):
    if body.truck_capacity_tons <= 0:
        raise HTTPException(status_code=422, detail="truck_capacity_tons must be positive")

    labels = [l for l in load_labels() if l["status"] in ("available_in_stock", "reserved")]

    if body.filters:
        if body.filters.warehouse_code:
            labels = [l for l in labels if l["warehouse_code"] == body.filters.warehouse_code]
        if body.filters.customer:
            labels = [l for l in labels if l["customer"] == body.filters.customer]

    return ffd(labels, body.truck_capacity_tons, body.max_iterations)
